# From: Intel

# Move gpio resource allocation from platform code to wifi driver

diff --git a/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c b/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c
index c7505d0..9bb934a 100644
--- a/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c
+++ b/mac80211/compat_wl12xx/drivers/net/wireless/wl12xx/sdio.c
@@ -230,6 +230,12 @@ static int __devinit wl1271_probe(struct sdio_func *func,
 		goto out_free_glue;
 	}
 
+	if (wlan_data->hw_init) {
+		ret = wlan_data->hw_init(wlan_data);
+		if (ret)
+			goto out_free_glue;
+	}
+
 	/* if sdio can keep power while host is suspended, enable wow */
 	mmcflags = sdio_get_host_pm_caps(func);
 	dev_dbg(glue->dev, "sdio PM caps = 0x%x\n", mmcflags);
@@ -292,12 +298,19 @@ out:
 static void __devexit wl1271_remove(struct sdio_func *func)
 {
 	struct wl12xx_sdio_glue *glue = sdio_get_drvdata(func);
-
+	struct wl12xx_platform_data *wlan_data;
+ 
+	wlan_data = wl12xx_get_platform_data();
+	BUG_ON(!wlan_data);
 	/* Undo decrement done above in wl1271_probe */
 	pm_runtime_get_noresume(&func->dev);
 
 	platform_device_del(glue->core);
 	platform_device_put(glue->core);
+	/* Free platform resources */
+	if (wlan_data->hw_deinit)
+		wlan_data->hw_deinit(wlan_data);
+
 	kfree(glue);
 }
 
diff --git a/mac80211/compat_wl12xx/include/linux/wl12xx.h b/mac80211/compat_wl12xx/include/linux/wl12xx.h
index 0d63731..4719d23 100644
--- a/mac80211/compat_wl12xx/include/linux/wl12xx.h
+++ b/mac80211/compat_wl12xx/include/linux/wl12xx.h
@@ -48,8 +48,13 @@ enum {
 
 struct wl12xx_platform_data {
 	void (*set_power)(bool enable);
+	int (*hw_init)(struct wl12xx_platform_data *pdata);
+	void (*hw_deinit)(struct wl12xx_platform_data *pdata);
 	/* SDIO only: IRQ number if WLAN_IRQ line is used, 0 for SDIO IRQs */
 	int irq;
+	/* gpio must be set to -EINVAL by platform code if gpio based irq is
+	not used */
+	int gpio;
 	bool use_eeprom;
 	int board_ref_clock;
 	int board_tcxo_clock;
